{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6b431fb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "import pandas as pd\n",
    "import sys\n",
    "import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "eaec0e9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#import the gene table, which has the NLRs for all accessions \n",
    "gene_table='/global/home/users/chandlersutherland/e16/Maize_NLRome_GeneTable.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6c9c1052",
   "metadata": {},
   "outputs": [],
   "source": [
    "#glob all the gff3 files across all genomes \n",
    "gff3_files=glob.glob('/global/scratch/users/chandlersutherland/e16/*/genome/*1.gff3')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0814356b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#define a function that takes a gff3 file path as input, and outputs an all gene bed file and an NLR specific bed file. \n",
    "def bed_generator(gff_filepath, gene_table):\n",
    "    #read in and cleanup gff3 file \n",
    "    gff3=pd.read_csv(gff_filepath, sep='\\t', skiprows=3, header=0, lineterminator='\\n')\n",
    "    gff3=gff3.reset_index()\n",
    "    gff3=gff3.rename(columns={gff3.columns[0]:'chr', gff3.columns[1]:'source', gff3.columns[2]:'sequence_ontology', gff3.columns[3]:'start', gff3.columns[4]:'end', gff3.columns[5]:'score', gff3.columns[6]:'strand', gff3.columns[7]:'phase', gff3.columns[8]:'attributes'})\n",
    "\n",
    "    #get metadata\n",
    "    accession=gff_filepath.split('/')[6]\n",
    "    basename=gff_filepath.split('/')[8].split('.')[0]\n",
    "    print(accession+ ' gff3 file loaded')\n",
    "    \n",
    "    #filter gff3 to just genes, and expand attributes  \n",
    "    genes = gff3[gff3['sequence_ontology'] == 'gene']\n",
    "    attributes=genes.iloc[:,8].str.split(';',expand=True)\n",
    "    genes.loc[:,'ID']=attributes.iloc[:,0].str.split('=', expand=True)[1]\n",
    "    genes.loc[:,'biotype']=attributes.iloc[:,1].str.split('=', expand=True)[1]\n",
    "    genes.loc[:,'logic_name']=attributes.iloc[:,2].str.split('=', expand=True)[1]\n",
    "    \n",
    "    #generate a bed file for all genes \n",
    "    bed = genes.loc[:,['chr', 'start', 'end', 'ID', 'strand']]\n",
    "    bed = bed.rename(columns={'chr':'chrom', 'start':'chromStart', 'end':'chromEnd', 'ID':'name', 'strand':'strand'})\n",
    "    bed.to_csv('/global/scratch/users/chandlersutherland/e16/' + accession + '/genome/' + basename + '_all_gene.bed', sep='\\t')\n",
    "    print('all gene bed written to '+'/global/scratch/users/chandlersutherland/e16/' + accession + '/genome/' + basename + '_all_gene.bed')\n",
    "    \n",
    "    #Load NLR gene names\n",
    "    gene=pd.read_csv(gene_table,sep = '\\t')\n",
    "    nlr_df=gene[gene['Ecotype'] == accession.lower()]\n",
    "    nlrs=nlr_df.loc[:,'Gene'].str.split('_', expand=True).iloc[:,0].str.replace('ZM', 'Zm').str.replace('AB', 'ab').unique()\n",
    "    \n",
    "    #quality stats \n",
    "    match=bed['name'].isin(nlrs).sum()\n",
    "    unmatch=len(nlrs)-match \n",
    "    print(str(match)+' NLRs found in gff file for '+accession+'. We are missing '+str(unmatch)+'.')\n",
    "    \n",
    "    #write out NLR bed file\n",
    "    nlr_bed=bed[bed['name'].isin(nlrs)]\n",
    "    nlr_bed.to_csv('/global/scratch/users/chandlersutherland/e16/' + accession + '/genome/' + basename + '_NLR.bed', sep='\\t')\n",
    "    print('NLR bed written to '+'/global/scratch/users/chandlersutherland/e16/' + accession + '/genome/' + basename + '_NLR.bed')\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "78276ad9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "B97 gff3 file loaded\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/global/software/sl-7.x86_64/modules/langs/python/3.7/lib/python3.7/site-packages/pandas/core/indexing.py:376: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self.obj[key] = _infer_fill_value(value)\n",
      "/global/software/sl-7.x86_64/modules/langs/python/3.7/lib/python3.7/site-packages/pandas/core/indexing.py:494: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self.obj[item] = s\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/B97/genome/Zm-B97-REFERENCE-NAM-1_all_gene.bed\n",
      "125 NLRs found in gff file for B97. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/B97/genome/Zm-B97-REFERENCE-NAM-1_NLR.bed\n",
      "Oh43 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/Oh43/genome/Zm-Oh43-REFERENCE-NAM-1_all_gene.bed\n",
      "131 NLRs found in gff file for Oh43. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/Oh43/genome/Zm-Oh43-REFERENCE-NAM-1_NLR.bed\n",
      "Tx303 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/Tx303/genome/Zm-Tx303-REFERENCE-NAM-1_all_gene.bed\n",
      "121 NLRs found in gff file for Tx303. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/Tx303/genome/Zm-Tx303-REFERENCE-NAM-1_NLR.bed\n",
      "CML69 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/CML69/genome/Zm-CML69-REFERENCE-NAM-1_all_gene.bed\n",
      "133 NLRs found in gff file for CML69. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/CML69/genome/Zm-CML69-REFERENCE-NAM-1_NLR.bed\n",
      "CML333 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/CML333/genome/Zm-CML333-REFERENCE-NAM-1_all_gene.bed\n",
      "131 NLRs found in gff file for CML333. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/CML333/genome/Zm-CML333-REFERENCE-NAM-1_NLR.bed\n",
      "NC358 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/NC358/genome/Zm-NC358-REFERENCE-NAM-1_all_gene.bed\n",
      "132 NLRs found in gff file for NC358. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/NC358/genome/Zm-NC358-REFERENCE-NAM-1_NLR.bed\n",
      "NC350 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/NC350/genome/Zm-NC350-REFERENCE-NAM-1_all_gene.bed\n",
      "136 NLRs found in gff file for NC350. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/NC350/genome/Zm-NC350-REFERENCE-NAM-1_NLR.bed\n",
      "Mo18W gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/Mo18W/genome/Zm-Mo18W-REFERENCE-NAM-1_all_gene.bed\n",
      "117 NLRs found in gff file for Mo18W. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/Mo18W/genome/Zm-Mo18W-REFERENCE-NAM-1_NLR.bed\n",
      "CML103 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/CML103/genome/Zm-CML103-REFERENCE-NAM-1_all_gene.bed\n",
      "119 NLRs found in gff file for CML103. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/CML103/genome/Zm-CML103-REFERENCE-NAM-1_NLR.bed\n",
      "CML247 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/CML247/genome/Zm-CML247-REFERENCE-NAM-1_all_gene.bed\n",
      "114 NLRs found in gff file for CML247. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/CML247/genome/Zm-CML247-REFERENCE-NAM-1_NLR.bed\n",
      "B73 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/B73/genome/Zm-B73-REFERENCE-NAM-5_all_gene.bed\n",
      "0 NLRs found in gff file for B73. We are missing 126.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/B73/genome/Zm-B73-REFERENCE-NAM-5_NLR.bed\n",
      "CML228 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/CML228/genome/Zm-CML228-REFERENCE-NAM-1_all_gene.bed\n",
      "126 NLRs found in gff file for CML228. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/CML228/genome/Zm-CML228-REFERENCE-NAM-1_NLR.bed\n",
      "P39 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/P39/genome/Zm-P39-REFERENCE-NAM-1_all_gene.bed\n",
      "118 NLRs found in gff file for P39. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/P39/genome/Zm-P39-REFERENCE-NAM-1_NLR.bed\n",
      "Oh7B gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/Oh7B/genome/Zm-Oh7B-REFERENCE-NAM-1_all_gene.bed\n",
      "121 NLRs found in gff file for Oh7B. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/Oh7B/genome/Zm-Oh7B-REFERENCE-NAM-1_NLR.bed\n",
      "MS71 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/MS71/genome/Zm-Ms71-REFERENCE-NAM-1_all_gene.bed\n",
      "122 NLRs found in gff file for MS71. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/MS71/genome/Zm-Ms71-REFERENCE-NAM-1_NLR.bed\n",
      "IL14H gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/IL14H/genome/Zm-Il14H-REFERENCE-NAM-1_all_gene.bed\n",
      "125 NLRs found in gff file for IL14H. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/IL14H/genome/Zm-Il14H-REFERENCE-NAM-1_NLR.bed\n",
      "Ki11 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/Ki11/genome/Zm-Ki11-REFERENCE-NAM-1_all_gene.bed\n",
      "120 NLRs found in gff file for Ki11. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/Ki11/genome/Zm-Ki11-REFERENCE-NAM-1_NLR.bed\n",
      "CML52 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/CML52/genome/Zm-CML52-REFERENCE-NAM-1_all_gene.bed\n",
      "130 NLRs found in gff file for CML52. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/CML52/genome/Zm-CML52-REFERENCE-NAM-1_NLR.bed\n",
      "CML322 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/CML322/genome/Zm-CML322-REFERENCE-NAM-1_all_gene.bed\n",
      "121 NLRs found in gff file for CML322. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/CML322/genome/Zm-CML322-REFERENCE-NAM-1_NLR.bed\n",
      "KY21 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/KY21/genome/Zm-Ky21-REFERENCE-NAM-1_all_gene.bed\n",
      "126 NLRs found in gff file for KY21. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/KY21/genome/Zm-Ky21-REFERENCE-NAM-1_NLR.bed\n",
      "M37W gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/M37W/genome/Zm-M37W-REFERENCE-NAM-1_all_gene.bed\n",
      "134 NLRs found in gff file for M37W. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/M37W/genome/Zm-M37W-REFERENCE-NAM-1_NLR.bed\n",
      "Tzi8 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/Tzi8/genome/Zm-Tzi8-REFERENCE-NAM-1_all_gene.bed\n",
      "130 NLRs found in gff file for Tzi8. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/Tzi8/genome/Zm-Tzi8-REFERENCE-NAM-1_NLR.bed\n",
      "M162W gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/M162W/genome/Zm-M162W-REFERENCE-NAM-1_all_gene.bed\n",
      "122 NLRs found in gff file for M162W. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/M162W/genome/Zm-M162W-REFERENCE-NAM-1_NLR.bed\n",
      "HP301 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/HP301/genome/Zm-HP301-REFERENCE-NAM-1_all_gene.bed\n",
      "116 NLRs found in gff file for HP301. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/HP301/genome/Zm-HP301-REFERENCE-NAM-1_NLR.bed\n",
      "CML277 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/CML277/genome/Zm-CML277-REFERENCE-NAM-1_all_gene.bed\n",
      "126 NLRs found in gff file for CML277. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/CML277/genome/Zm-CML277-REFERENCE-NAM-1_NLR.bed\n",
      "Ki3 gff3 file loaded\n",
      "all gene bed written to /global/scratch/users/chandlersutherland/e16/Ki3/genome/Zm-Ki3-REFERENCE-NAM-1_all_gene.bed\n",
      "129 NLRs found in gff file for Ki3. We are missing 0.\n",
      "NLR bed written to /global/scratch/users/chandlersutherland/e16/Ki3/genome/Zm-Ki3-REFERENCE-NAM-1_NLR.bed\n"
     ]
    }
   ],
   "source": [
    "for i in gff3_files:\n",
    "    bed_generator(str(i), gene_table)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.7",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
