---
title: "fig_generation"
author: "Chandler Sutherland"
date: "2023-11-17"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Generating the plots shown in Fig3 and Supplemental Figure 3
```{r}
library(data.table)
library(ggplot2)
library(tidyverse)
library(ggsignif)
library(ggpubr)
library(ggbeeswarm)
library(pheatmap)
library(patchwork)
library(introdataviz)
library(viridis)
```

Fig 3a: transcriptome PCA 
Read in PC scores generated by the pca_transcriptome.R file, plot PC1 and PC2, then color by tissue type of origin. 
```{r}
#Read in PC scoreswithout endosperm and embryo 
pc_scores2 <- read.csv('//wsl.localhost//Ubuntu//home//chandlersutherland//e16_scratch//PC2_tpm_matrix.csv') 

pc_plot2 <- pc_scores2 %>% 
  # create the plot
  ggplot(aes(x = PC1, y = PC2, color=tissue)) +
  geom_point(size=0.5)+
  xlab('PC1 (21.6% of variance)')+ #pulled from eigenvalues calculated in pca_transcriptome.R
  ylab('PC2 (11.3% of variance)')+
  theme_classic(base_size=10)+
  scale_color_viridis(discrete=TRUE, option="turbo")+
  theme(text=element_text(size=10))+
  theme(legend.key.height=unit(.65,"line"))+
  guides(fill = guide_legend(byrow = TRUE))+
  theme(legend.key.size = unit(.25, "cm"), 
        legend.title=element_blank(), 
        legend.spacing.y = unit(0, "cm"), 
        legend.spacing.x=unit(0.1, 'cm'), 
        legend.margin=margin(0,0,0,-1), 
        legend.box.margin=margin(t=0, r=0, b=0, l=-5))

ggsave('C://Users//chand//Box Sync//Krasileva_Lab//Research//chandler//Krasileva Lab//E16//figure panels//pca2.png', plot=pc_plot2, dpi='retina', width=3, height=2.1)

#save per tissue color codes for other plots 
tissue_colors <- scales::viridis_pal(option='turbo')(length(unique(pc_scores2$tissue)))
t_list <- unique(pc_scores2$tissue) %>% sort()
tissue_df <- as.data.frame(t_list, tissue_colors) %>% rownames_to_column('tissue_color')

pc_plot2
```

Figure 3B: split violin plots comparing hv and non-hvNLR aggregate expression across tissue types
```{r}
#remove unmappable NLRs 
unmappable <- c('Zm00039ab351270', 'Zm00026ab135540', 'Zm00036ab418650',
                'Zm00001eb091500', 'Zm00001eb164880', 'Zm00001eb343890',
                'Zm00001eb391100', 'Zm00001eb405870', 'Zm00001eb405900',
                'Zm00001eb405930', 'Zm00033ab429000', 'Zm00029ab367660')

#import gene information 
gene_table <- read.csv('Maize_NLRome_GeneTable.txt', sep='\t') %>% subset(select=c('Gene', 'Ecotype', 'HV', 'Clade'))
gene_table$Ecotype <- gene_table$Ecotype %>% toupper()
gene_table$Gene <- gene_table$Gene %>% 
  str_replace('ZM', 'Zm') %>% 
  str_replace('AB', 'ab') %>% 
  str_replace("EB", 'eb') %>% 
  str_replace('_P001', '')

#import nlr TPM table, average biological replicates 
all_tpm <- read.csv('//wsl.localhost//Ubuntu//home//chandlersutherland//e16_scratch//all_nlr_tissue.csv') %>% 
  filter(!tissue %in% c('endosperm', 'embryo')) %>% 
  filter(!name %in% unmappable) %>% 
  merge(gene_table, by.x='name', by.y='Gene')

all_tpm_avg <- all_tpm %>%
  group_by(Clade, HV, accession, tissue, name, chrom, chromStart, chromEnd, strand) %>%
  summarize(log2_TPM=mean(log2.TPM., na.rm = T)) %>%
  ungroup() 

#calculate significance of difference, correct by Benjamani-Hochberg 
p_tbl <- compare_means(log2_TPM~HV, data=all_tpm_avg, group.by='tissue', p.adjust.method = 'BH') %>% arrange(p.adj)
order_p <- p_tbl %>% pull(tissue)
p_annot <- p_tbl %>% pull(p.signif)

#clean up data for pretty plotting 
all_tpm_avg$tissue <- factor(all_tpm_avg$tissue, levels=c(rev(order_p)))
all_tpm_avg <- all_tpm_avg %>% mutate(HV=case_match(HV, 0 ~ "non-hv", 1~"hv")) #I don't know why this takes so long.. 
all_tpm_avg$HV <- factor(all_tpm_avg$HV, levels=c('non-hv', 'hv'))
```

```{r}
#plot 
hv_by_tissue <- ggplot(all_tpm_avg) +
  introdataviz::geom_split_violin(aes(x = tissue, y = log2_TPM, fill = HV), 
                                  alpha=0.4, trim = TRUE, scale="width")+
  stat_summary(mapping=aes(x = tissue, y = log2_TPM, fill=HV),
               fun = median, geom = "point", show.legend = F, shape=21, color='black',
               position = position_dodge(.175)) +
  scale_fill_manual(values=c('#00BFC4',  '#F8766D'))+
  scale_color_manual(values=c('#00BFC4',  '#F8766D'))+
  scale_y_continuous(limits = c(0,13), expand = c(0, 0)) +
  geom_signif(stat='identity', 
                      data=data.frame(x = c(0.75, 1.75, 2.75, 3.75, 4.75, 5.75, 6.75, 7.75),
                     xend=c(1.25, 2.25, 3.25, 4.25, 5.25, 6.25, 7.25, 8.25),
                      y = c(12, 12, 12, 12, 12, 12, 12, 12),
                    annotation2 = c('*', ' ** ', '**', '****', ' **** ', '  ****  ', '   ****   ', '    ****    ')), 
              aes(x=x, xend=xend, y=y, yend=y, annotation=annotation2, textsize=4))+
  theme_classic(base_size = 10)+
  ylab('log2(TPM)')+
  xlab('')+
  theme(legend.key.size = unit(.25, "cm"), 
        legend.title=element_blank(), 
        legend.position='bottom', 
        text = element_text(size=10), 
        legend.spacing.y = unit(0, "cm"), 
        legend.spacing.x=unit(0.1, 'cm'), 
        legend.margin=margin(-3,0,0,0), 
        legend.box.margin=margin(-12,-12,0,0), 
        axis.text.y=element_text(size=10))

ggsave('C://Users//chand//Box Sync//Krasileva_Lab//Research//chandler//Krasileva Lab//E16//figure panels//rna_hv_tissue2.png', plot=hv_by_tissue, dpi='retina', width=3.5, height=2.1)

hv_by_tissue
```
```{r}
all_tpm_avg %>% group_by(HV, tissue) %>% summarize(n=n())
```



Figure 3C: per clade heatmap of expression 
```{r}
#reformat tpm table 
#define a function that applies a sample ID based on the replicates
reformat <- function(split_df){
  #by_tissue <- split(split_df, split_df$tissue)
  split_df <- split_df  %>% group_by(rep) %>% mutate(sample_id=cur_group_id()) %>% mutate(tissue=paste(tissue, as.character(sample_id), sep='_'))
  split_df 
}

#define a function that applies reformat to each accession data frame 
by_tissue <- function(by_accession){
  one_acc <- lapply(split(by_accession, by_accession$tissue), reformat) %>% bind_rows()
  one_acc
}

#bring it all together 
all <- all_tpm  %>%  split(all_tpm$accession)
reformed <- lapply(all, by_tissue) %>% bind_rows() %>% subset(select=c(HV, accession, name, tissue, log2.TPM.)) %>%
  unique() %>%
  pivot_wider(names_from=tissue, values_from=log2.TPM.) %>% 
  remove_rownames() %>% 
  column_to_rownames('name')

#create an hv expression matrix 
p <- reformed %>% 
  filter(HV==1) %>% 
  subset(select=-c(HV, accession, tip_3, middle_3, root_3, base_3)) %>% #remove extra bioreps 
  filter_all(any_vars(. != 0))  %>% #remove genes with no variation across tissue types (would also remove all 0 genes)
  subset(select=c('root_1', 'root_2', 'ear_1', 'ear_2', 'shoot_1', 'shoot_2', 'base_1', 'base_2', 'tassel_1', 'tassel_2',
                  'anther_1', 'anther_2', 
                   'middle_1', 'middle_2', 'tip_1', 'tip_2'))
```



To define the gene expression categories, we compared TPM values across the 10 tissues (leaf tip, leaf middle, leaf base, root, shoot, ear, anther, tassel, endosperm, and embryo). We defined tissue-specific expression as TPM ≥ 1 in at least 1 tissue and TPM < 1 in at least 1 tissue, constitutive expression as TPM ≥ 1 in all 10 tissues, and silent as TPM < 1 in all 10 tissues. 
```{r}
by_category <- all_tpm %>% 
  group_by(accession, name, tissue, HV, chrom, chromStart,chromEnd, strand) %>%
  summarize(TPM=mean(TPM, na.rm=TRUE)) %>%
  mutate(HV=case_match(HV, 0 ~ "non-hv", 1~"hv")) %>%
  pivot_wider(names_from=tissue, values_from=TPM) %>%
  mutate(expr_category=case_when((
    (
      (anther >= 1) | (base >= 1) | (ear >= 1) | (middle >= 1) | (root >= 1) | (shoot >= 1) | (tip >= 1) | (tassel >=1)
      ) & 
      (
        (anther < 1) | (base < 1) | (ear < 1) | (middle < 1) | (root < 1) | (shoot < 1) | (tip < 1) | (tassel < 1)
        )
    )  ~'tissue_specific',
    (
      (anther >= 1) & (base >= 1) & (ear >= 1) & (middle >= 1) & (root >= 1) & (shoot >= 1) & (tip >= 1) & (tassel >=1)
      ) ~ 'constitutive',
    (
      (anther < 1) & (base < 1) & (ear < 1) & (middle < 1) & (root < 1) & (shoot < 1) & (tip < 1) & (tassel < 1)
      ) ~ 'silent'))


by_category$HV <- factor(by_category$HV, levels=c('non-hv','hv'))


all_tpm_avg <- by_category %>% subset(select=c(name, expr_category)) %>% unique() %>% merge(all_tpm_avg) %>% 
  mutate(Clade_adj=recode(Clade, 'Int3480_75_130_L_68'='Int3480', 
                          'Int4787_129_172_L_83'='RppM-like', 
                          'Int6329_131_253_L_122_209_L_80_142_L_61'='RppC-like', 'Int6329_131_253_L_122_209_R_35_43_R_23'='RppC-like',
                          'Int6329_131_253_L_122_209_L_80_142_R_18'='RppC-like',
                          'Int6648_150_178_L_128_144_R_93'='Rp1-like'))

write_csv(all_tpm_avg, file="C://Users//chand//Box Sync//Krasileva_Lab//Research//chandler//Krasileva Lab//E16//intermediate_data//nlr_tpm.csv")

coordinates <- all_tpm_avg %>% 
  subset(select=c('accession', 'name', 'Clade_adj', 'HV', 'expr_category', 'chrom', 'chromStart', 'chromEnd', 'strand')) %>% 
  unique()

write_csv(coordinates, file="C://Users//chand//Box Sync//Krasileva_Lab//Research//chandler//Krasileva Lab//E16//intermediate_data//nlr_coords.csv")
```

```{r}
#create an annotation matrix 

#import subpopulation membership 
subpopulations <- read_table('//wsl.localhost//Ubuntu//home//chandlersutherland//e16//nam_genome_info.txt', col_names=c('Assembly', 'Grin', 'accession_id', 'source', 'cross_reference', 'subpopulation', 'stock'), skip=1) %>% separate(Assembly, sep='-', into=c(NA, 'accession', NA, NA, NA))
subpopulations$accession <- subpopulations$accession %>% toupper()
subpopulations <- subpopulations %>% subset(select=c('accession', 'subpopulation'))
subpopulations[nrow(subpopulations) + 1,] = list('B73', 'Stiff stalk')
subpopulations <- subpopulations %>% mutate(subpopulation =recode(subpopulation, 'Temporate/tropical'='Temporate/Tropical'))%>% 
  mutate(subpopulation =recode(subpopulation, 'Temporate/Tropical'='Mixed', 'Non-stiff-stalk'='Non-stiff stalk', 'Sweet'='Sweetcorn'))
subpopulations$subpopulation <- factor(subpopulations$subpopulation, levels=c('Stiff stalk', 'Non-stiff stalk', 'Mixed', 'Popcorn', 'Sweetcorn', 'Tropical'))


#apply clade membership from geneTable
row_annotation <- all_tpm_avg %>%
  filter(HV=='hv') %>% 
  subset(select=c(name, accession, Clade_adj, expr_category)) %>% 
  unique() %>% 
  merge(subpopulations) %>% 
  subset(select=c(name,  expr_category, subpopulation, Clade_adj)) %>% 
  remove_rownames() %>% 
  column_to_rownames('name')

#create column annotations
samples <- colnames(p)
col_annotation <- as.data.frame(samples) %>% separate(samples, '_', into=c('tissue', NA), remove=F) %>% column_to_rownames('samples')

#create color annotation list 
my_col = list(
  Clade_adj=c('Int3480'='#fee090', 
                              'RppM-like'='#d73027', 
                              'RppC-like'='#fc8d59', 
                              'Rp1-like'='#4575b4'), 
  subpopulation=c('Stiff stalk'='#ffe3a9', 
                              'Non-stiff stalk'='#80d6f6', 
                              'Mixed'='#d7d7d7', 
                              'Popcorn'='#e6b1ff', 
                              'Sweetcorn'='#ffbf70', 
                              'Tropical'='#78d5a0'),
  expr_category=c('tissue_specific'='#00BA38', 
                  'silent'='#F8766D',
                  'constitutive'='#619CFF'
                  ),
  tissue=c('anther'='#30123BFF', 
           'base'='#4777EFFF', 
           'ear'='#1BD0D5FF', 
           'middle'='#62FC6BFF', 
           'root'='#D2E935FF', 
           'shoot'='#FE9B2DFF', 
           'tassel'='#DB3A07FF', 
           'tip'='#7A0403FF'))

#get gene lists for each hv clade
Int3480 <- row_annotation %>% filter(Clade_adj=='Int3480') %>% rownames()
Rp1 <- row_annotation %>% filter(Clade_adj=='Rp1-like') %>% rownames()
RppM <- row_annotation %>% filter(Clade_adj=='RppM-like') %>% rownames()
RppC <- row_annotation %>% filter(Clade_adj=='RppC-like') %>% rownames()

#standardize mat breaks to I can create each heatmap individually 
mat_breaks <- seq(min(p, na.rm=TRUE), max(p, na.rm=TRUE), length.out = 100)
```

```{r}
#plot! 
Int3480_m <- p[rownames(p) %in% Int3480,]
pheatmap(Int3480_m, 
         cluster_cols=FALSE,
         show_rownames=FALSE,
         show_colnames=FALSE,
         annotation_row=row_annotation, 
         annotation_colors=my_col, 
         cellheight=1, cellwidth=5, 
         border_color=NA,
         legend=FALSE,
         annotation_legend=FALSE, 
         annotation_names_row=FALSE,
         treeheight_row=30,
         breaks=mat_breaks, 
        filename='C://Users//chand//Box Sync//Krasileva_Lab//Research//chandler//Krasileva Lab//E16//figure panels//Int3480_2.png',
        color=inferno(100))

Rp1_m <- p[rownames(p) %in% Rp1,]
pheatmap(Rp1_m, 
         cluster_cols=FALSE, 
         show_rownames=FALSE, 
         show_colnames=FALSE,
         annotation_row=row_annotation, 
         annotation_colors=my_col, 
         cellheight=1, cellwidth=5, 
         legend=FALSE,
         annotation_legend=FALSE, 
         annotation_names_row=FALSE,
         border_color=NA,
         breaks=mat_breaks, 
         color=inferno(100),
         filename='C://Users//chand//Box Sync//Krasileva_Lab//Research//chandler//Krasileva Lab//E16//figure panels//Rp1_2.png',
         treeheight_row=30
         )

RppM_m <- p[rownames(p) %in% RppM,]
pheatmap(RppM_m, 
         cluster_cols=FALSE, 
         show_rownames=FALSE, 
         show_colnames=FALSE,
         annotation_row=row_annotation, 
         annotation_colors=my_col, 
         cellheight=1, cellwidth=5, 
         legend=FALSE,
         annotation_legend=FALSE, 
         annotation_names_row=FALSE,
         annotation_names_col=FALSE,
         border_color=NA,
         annotation_col=col_annotation, #as the leftmost plot it gets the color annotation 
         breaks=mat_breaks, 
         color=inferno(100), 
         filename='C://Users//chand//Box Sync//Krasileva_Lab//Research//chandler//Krasileva Lab//E16//figure panels//RppM_2.png',
         treeheight_row=30
         )


RppC_m <- p[rownames(p) %in% RppC,]
pheatmap(RppC_m, 
         cluster_cols=FALSE, 
         show_rownames=FALSE, 
         annotation_row=row_annotation, 
         annotation_colors=my_col, 
         cellheight=1, cellwidth=5, 
         legend=FALSE,
         annotation_legend=FALSE, 
         annotation_names_row=FALSE,
         breaks=mat_breaks, 
         color=inferno(100),
         show_colnames = FALSE,
         treeheight_row=30,
         filename='C://Users//chand//Box Sync//Krasileva_Lab//Research//chandler//Krasileva Lab//E16//figure panels//RppC_2.png',
         border_color=NA)

#using this for the legend 
pheatmap(RppC_m, 
         cluster_cols=FALSE, 
         show_rownames=FALSE, 
         annotation_row=row_annotation, 
         annotation_colors=my_col, 
         cellheight=1, cellwidth=5, 
         legend=TRUE,
         annotation_legend=TRUE, 
         annotation_names_row=FALSE,
         breaks=mat_breaks, 
         color=inferno(100),
         show_colnames = TRUE,
         treeheight_row=30,
         filename='C://Users//chand//Box Sync//Krasileva_Lab//Research//chandler//Krasileva Lab//E16//figure panels//legend.png',
                 border_color=NA
         )
```

```{r}
Int3480_m
#Zm00001eb135110 tassel 1 and 2
#Zm00021ab135990 tassel 2
#Zm00019ab129030	and Zm00019ab129050 tassel 2 
big_df <- lapply(all, by_tissue) %>% bind_rows() %>% subset(select=c(HV, accession, name, tissue, log2.TPM.)) %>% pivot_wider(names_from=tissue, values_from=log2.TPM.)

big_df %>% group_by(accession) %>% summarize(anther_1=sum(is.na(anther_1)), 
                                             anther_2=sum(is.na(anther_2)), 
                                             base_1=sum(is.na(base_1)), 
                                             base_2=sum(is.na(base_2)),
                                             ear_1=sum(is.na(ear_1)), 
                                             ear_2=sum(is.na(ear_2)),
                                             middle_1=sum(is.na(middle_1)), 
                                             middle_2=sum(is.na(middle_2)),
                                             root_1=sum(is.na(root_1)), 
                                             root_2=sum(is.na(root_2)),
                                             shoot_1=sum(is.na(shoot_1)), 
                                             shoot_2=sum(is.na(shoot_2)),
                                             tip_1=sum(is.na(tip_1)), 
                                             tip_2=sum(is.na(tip_2)),
                                             tassel_1=sum(is.na(tassel_1)), 
                                             tassel_2=sum(is.na(tassel_2)))
```

```{r}
tpm_with_clade %>% filter(HV.x=='hv') %>% group_by(Clade_adj) %>% summarize(median=median(log2_TPM))
tpm_with_clade %>% group_by(tissue) %>% summarize(median=median(log2_TPM)) %>% arrange(median)
```



```{r}
by_category %>% group_by(HV, accession, expr_category) %>% summarize(n=n()) %>% mutate(prop=n/sum(n)*100) %>%
  filter(!is.na(expr_category))%>%
  compare_means(prop~HV, group.by='expr_category', p.adjust.method='BH', data=., method='t.test') %>% arrange(p.adj)


compare_means(middle~HV, data=by_category, group.by='expr_category', p.adjust.method = 'BH') %>% arrange(p.adj)

by_category %>% 
  subset(select=c(accession, name, expr_category, HV)) %>% 
  merge(all_tpm_avg, by=c('accession', 'name')) %>%
  filter(!is.na(expr_category))%>%
  ggplot() +
  introdataviz::geom_split_violin(aes(x = tissue, y = log2_TPM, fill = HV), 
                                  alpha=0.4, trim = TRUE, scale="width")+
  stat_summary(mapping=aes(x = tissue, y = log2_TPM, fill=HV),
               fun = median, geom = "point", show.legend = F, shape=21, color='black',
               position = position_dodge(.175)) +
  scale_fill_manual(values=c('#00BFC4',  '#F8766D'))+
  scale_color_manual(values=c('#00BFC4',  '#F8766D'))+
  scale_y_continuous(limits = c(0,13), expand = c(0, 0)) +
  theme_classic(base_size = 10)+
  ylab('log2(TPM)')+
  xlab('')+
  theme(legend.key.size = unit(.25, "cm"), 
        legend.title=element_blank(), 
        legend.position='bottom', 
        text = element_text(size=10), 
        legend.spacing.y = unit(0, "cm"), 
        legend.spacing.x=unit(0.1, 'cm'), 
        legend.margin=margin(-3,0,0,0), 
        legend.box.margin=margin(-12,-12,0,0), 
        axis.text.y=element_text(size=10))+
  facet_wrap(~expr_category)
```


```{r}
cat_plot <- all_tpm_avg %>% 
  group_by(HV, Clade_adj, expr_category) %>% summarize(n=n()) %>% mutate(prop=n/sum(n)*100)

cat_plot$expr_category <- factor(cat_plot$expr_category, levels=c('silent', 'tissue_specific', 'constitutive'))  

by_category_plot <- cat_plot %>% 
  group_by(HV, expr_category) %>% 
  mutate(rank=rank(prop)) %>%
  mutate(Clade_adj=reorder(Clade_adj, -rank)) %>%
  ungroup() %>%
  ggplot(aes(x=Clade_adj, y=prop, fill=expr_category))+
  geom_col()+
  facet_grid(~HV, scales='free', space='free')+
  scale_fill_manual(values=c('silent'='#00C59E', 'tissue_specific'='#5977E7', 'constitutive'='#E56F35'))+
  ylab('% of Clade')+
  xlab('Clade')+
  #labs(fill='Gene Expression Category')+
  scale_fill_discrete(name = "Gene Expression Category", labels = c("Silent", "Tissue Specific", "Constitutive"))+
  theme_classic()+
  theme(axis.text.x = element_blank(),
        legend.position = 'bottom',
        text = element_text(size=10)
        #axis.ticks.x = element_blank()
        )

ggsave(by_category_plot, filename='C://Users//chand//Box Sync//Krasileva_Lab//Research//chandler//Krasileva Lab//E16//figure panels//expression_by_category.png', width=6.5, height=2.5, units='in')

```


